### PR TITLE
Full documentation within the smart contract

### PR DESCRIPTION
**Pull request by JonathonDunford**
Closes: https://github.com/forkdelta/smart_contract/issues/16

Full documentation within the smart contract utilizing Doxygen format as outlined by Solidity: http://solidity.readthedocs.io/en/v0.3.2/layout-of-source-files.html

### CODE REVIEW

File: contracts/ForkDelta.sol [hunk 1]
```diff
@@ -23,11 +27,13 @@ contract ForkDelta {
   event Deposit(address token, address user, uint amount, uint balance);
   event Withdraw(address token, address user, uint amount, uint balance);
 
+  /// This is a modifier for functions to check if the sending user address is the same as the admin user address.
   modifier isAdmin() {
       require(msg.sender == admin);
       _;
   }
 
+  /// Initialization function. This is only called on contract creation.
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: I would call this the "constructor".

File: contracts/ForkDelta.sol [hunk 2]
```diff
@@ -37,46 +43,71 @@ contract ForkDelta {
     depositingTokenFlag = false;
   }
 
+  /// This default function prevents people sending Ether directly into the contract.
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: I would change this to:
/// The fallback function. Ether transfered into the contract is not accepted.

File: contracts/ForkDelta.sol [hunk 3]
```diff
@@ -37,46 +43,71 @@ contract ForkDelta {
...
+  */
   function withdraw(uint amount) public {
     require(tokens[0][msg.sender] >= amount);
     tokens[0][msg.sender] = tokens[0][msg.sender].sub(amount);
     msg.sender.transfer(amount);
     emit Withdraw(0, msg.sender, amount, tokens[0][msg.sender]);
   }
 
+  /**
+  * IMPORTANT: Remember to call Token(address).approve(this, amount) or this contract will not be able to do the transfer on your behalf.
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: I would not make this note so prominent. The approve is on our UI code and that's about it.

File: contracts/ForkDelta.sol [hunk 4]
```diff
@@ -105,18 +151,54 @@ contract ForkDelta {
     emit Withdraw(token, msg.sender, amount, tokens[token][msg.sender]);
   }
 
+  /**
+  * Retrieves the balance of a token based on a user address and token address.
+  * @param token Ethereum contract address of the token
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: @param token Ethereum contract address of the token or 0 for ether

File: contracts/ForkDelta.sol [hunk 5]
```diff
@@ -105,18 +151,54 @@ contract ForkDelta {
...
+  * @param token Ethereum contract address of the token 
+  * @param user Ethereum address of the user
+  * @return the amount of tokens on the exchange for a given user address
+  */
   function balanceOf(address token, address user) public constant returns (uint) {
     return tokens[token][user];
   }
 
+  /**
+  * Note: tokenGet & tokenGive can be the Ethereum contract address.
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: or 0 for ether

File: contracts/ForkDelta.sol [hunk 6]
```diff
@@ -105,18 +151,54 @@ contract ForkDelta {
...
+  * @param nonce arbitrary random number
+  */
   function order(address tokenGet, uint amountGet, address tokenGive, uint amountGive, uint expires, uint nonce) public {
     bytes32 hash = sha256(this, tokenGet, amountGet, tokenGive, amountGive, expires, nonce);
     orders[msg.sender][hash] = true;
     emit Order(tokenGet, amountGet, tokenGive, amountGive, expires, nonce, msg.sender);
   }
 
+  /**
+  * Note: tokenGet & tokenGive can be the Ethereum contract address.
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Can we generally have the notes below the definition of what the function does?

File: contracts/ForkDelta.sol [hunk 7]
```diff
@@ -128,10 +210,25 @@ contract ForkDelta {
     emit Trade(tokenGet, amount, tokenGive, amountGive.mul(amount).div(amountGet), user, msg.sender);
   }
 
+  /**
+  * Note: tokenGet & tokenGive can be the Ethereum contract address.
+  * Note: amount is in amountGet / tokenGet terms.
+  * This is a private function and can only be called from trade().
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: * This is a private function and is only being called from trade().

File: contracts/ForkDelta.sol [hunk 8]
```diff
@@ -128,10 +210,25 @@ contract ForkDelta {
...
+  * Handles the movement of funds when a trade occurs.
+  * Takes fees.
+  * Updates token balances for both sender and receiver
```
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: sender, receiver => buyer and seller

File: contracts/Token.sol [hunk 9]
```diff
@@ -1,5 +1,9 @@
 pragma solidity ^0.4.21;
 
+/**
+ * @title Token
+ * @dev Token functionality necessary for working with tokens within the exchange contract.
```
Comment (mmhh1910) [inline contracts/Token.sol]: Token interface necessary for working with tokens within the exchange contract.

**Review by mmhh1910 — APPROVED** [review]
Great work! I've added a bunch of comments. Please implement them as you see fit. Approving anyway.

File: contracts/ForkDelta.sol (hunk 6 above)
Comment (JonathonDunford) [reply contracts/ForkDelta.sol]: Yeah, makes more sense that way.

**Comment by JonathonDunford** [comment]
Thanks @betterdelta 

All of your suggestions were reasonable and pointed out some things I missed. Updated to reflect. Will merge now.

**Pull request merged**